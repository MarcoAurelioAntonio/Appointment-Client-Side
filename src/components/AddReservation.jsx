import React, { useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useLocation } from 'react-router';
import { postReservationToAPI, reset } from '../redux/reservationsSlice';
import NavMenu from './NavMenu';

const AddReservation = () => {
  // Retrieve data from other location
  const currentLocation = useLocation();
  const courseSelected = currentLocation.state || null;
  const [course, setCourse] = useState(courseSelected?.id);
  const { status, error, enrolled } = useSelector(
    (store) => store.reservations,
  );
  const user = useSelector((store) => store.users.current_user);
  const courses = useSelector((store) => store.courses.courses);
  const dispatch = useDispatch();
  // const history = useNavigate();

  const handleChange = (ev) => {
    setCourse(ev.target.value);
  };
  const handleClick = (ev) => {
    ev.preventDefault();
    dispatch(reset());
    const reservation = {
      user_id: user.id,
      course_id: course,
    };
    dispatch(postReservationToAPI(reservation));
    setTimeout(() => dispatch(reset()), 5000);
  };

  return (
    <div className="flex">
      <NavMenu bgColor="gray" isHide isBacking={false} />
      <section className="enrollment-content">

        <div className="enrollment-title">
          Are you ready to unlock your full potential?
        </div>
        <p className="enrollment-text">
          By investing in your education, you open doors to new knowledge, skills,
          and opportunities. Embrace the chance to broaden your horizons, gain
          expertise, and boost your career prospects. Remember, learning is a
          lifelong journey, and each course you undertake is a valuable milestone
          along the way. Take that leap of faith and embark on an exciting
          educational adventure today!
        </p>
        <div className="enroll-container">
          <form onSubmit={handleClick}>
            <label htmlFor="name">
              <input
                className="enroll-field"
                type="text"
                placeholder="Enter Username"
                value={user.name}
                name="name"
                disabled
              />
            </label>
            <br />
            <label htmlFor="Course">
              {courseSelected ? (
                <input
                  className="enroll-field"
                  type="text"
                  value={courseSelected.name}
                  name="name"
                  disabled
                />
              ) : (
                <select
                  className="enroll-field"
                  name="course"
                  onChange={handleChange}
                  required
                >
                  <option value="">Select your course</option>
                  {courses.map((item) => (
                    <option key={item.id} value={item.id}>
                      {item.name}
                    </option>
                  ))}
                </select>
              )}
            </label>
            <button
              className="enroll-btn-ken"
              type="submit"
              disabled={status === 'loading' && true}
            >
              ENROLL NOW
            </button>
          </form>
        </div>
        <section className="msg-section">
          <p className="error">{error}</p>
          <p className="success">
            {enrolled && 'You have been enrolled successfully. Enjoy learning!'}
          </p>
        </section>
      </section>
    </div>
  );
};

export default AddReservation;
