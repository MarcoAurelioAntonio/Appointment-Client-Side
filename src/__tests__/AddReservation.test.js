import { render, screen } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import configureStore from 'redux-mock-store';
import AddReservation from '../components/AddReservation';

it('Should display 1 button on the screen', () => {
  const mockStore = configureStore([]);
  const user = {
    id: 1,
    name: 'Teddy',
  };

  const courses = [
    {
      id: 1,
      name: 'Master React',
      description: 'Lorem ipsu dolor',
    },
  ];

  const initialState = {
    users: {
      current_user: user,
      status: 'idle',
      loading: false,
      error: null,
    },
    courses: {
      courses,
      course: {},
      status: 'idle',
      loading: false,
      error: null,
      deleted: false,
    },
    reservations: {
      reservations: [],
      status: 'idle',
      loading: false,
      error: null,
      userId: null,
      enrolled: false,
    },
  };

  const store = mockStore(initialState);
  render(
    <BrowserRouter>
      <Provider store={store}>
        <AddReservation />
      </Provider>
    </BrowserRouter>,
  );
  const buttons = screen.getAllByText('ENROLL NOW');
  const courseName = screen.getByText('Master React');

  expect(buttons).toHaveLength(1);
  expect(courseName).toContainHTML(courses[0].name);
  expect(store.getState()).toMatchSnapshot();
});
