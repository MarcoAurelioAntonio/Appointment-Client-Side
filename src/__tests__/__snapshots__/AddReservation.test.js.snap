// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should display 1 button on the screen 1`] = `
Object {
  "courses": Object {
    "course": Object {},
    "courses": Array [
      Object {
        "description": "Lorem ipsu dolor",
        "id": 1,
        "name": "Master React",
      },
    ],
    "deleted": false,
    "error": null,
    "loading": false,
    "status": "idle",
  },
  "reservations": Object {
    "enrolled": false,
    "error": null,
    "loading": false,
    "reservations": Array [],
    "status": "idle",
    "userId": null,
  },
  "users": Object {
    "current_user": Object {
      "id": 1,
      "name": "Teddy",
    },
    "error": null,
    "loading": false,
    "status": "idle",
  },
}
`;
